encoding/video-artifacts.md

```diff
@@ -232,6 +234,75 @@ or a manual filtering of dehalo\_alpha with dhhmask
 
 [![field-noise.jpg](images/3cnvimage103.png)](https://diff.pics/84URvW5IYSdO/1)
 
+TODO
+
+
+### Underflow / Overflow
+
+While most of the anime produced
+use the YUV 8-bit limited range[^4],
+we occasionally find some videos having the "limited range" flag set
+while containing full range content.
+This often results in oversaturated colors and weird brightness.
+Thus, it is strongly recommended
+to check the brightness levels
+of the 8-bit source with `hist.Levels()`.
+
+[![Example of underflow (click for comparison)](images/underflow.jpg)](https://slowpics.org/comparison/f125e799-fdff-4c4c-8c9d-707af021bd88)
+
+[![Example of overflow (click for comparison)](images/overflow.jpg)](https://slowpics.org/comparison/6e24ffe9-e068-4f33-b2e7-639031d512f2)
+
+To fix this problem,
+simply use [`std.Levels()`][std-Levels] like so:
+
+```py
+# This only applies to 8 bit clips!
+clip = clip.std.Levels(min_in=0, max_in=255, min_out=16, max_out=235, planes=0)      # y plane
+clip = clip.std.Levels(min_in=0, max_in=255, min_out=16, max_out=240, planes=[1,2])  # u&v planes
+```
+
+or set the "full range" flag on the video,
+so the values can be interpreted accordingly.
+Limited range video is more widely supported
+and players may ignore the "full range" flag,
+which results in interpreting full range content 
+in a limited context.
+
+[std-Levels]: http://www.vapoursynth.com/doc/functions/levels.html
+
+Because limited precision with only 8 bit per channel
+may lead to rounding errors quickly,
+we prefer adjusting the levels
+(and our filtering in general)
+with higher precision, 
+such as 16 bit or float (32 bit).
+For these clips,
+you would use the following[^5]:
+
+```py
+# 16 bit
+clip = clip.std.Levels(min_in=0, max_in=255 * 2**8, min_out=16 * 2**8, max_out=235 * 2**8, planes=0)      # y plane
+clip = clip.std.Levels(min_in=0, max_in=255 * 2**8, min_out=16 * 2**8, max_out=240 * 2**8, planes=[1,2])  # u&v planes
```

Comment:
Bitshifting doesn't work for full range, since 255 << 8 = 65280 ≠ 65535. `max_in=2**16 - 1` would be correct.

Reply:
Right, nice find. Also good point about bit shifting. Because I use that operator so rarely, I forgot about it.